h/presenters/annotation_json.py

```diff
@@ -60,10 +58,14 @@ def _get_read_permission(self):
             # It's not shared so only the owner can read it
             return self.annotation.userid
 
-        if security.Everyone in principals_allowed_by_permission(
-            AnnotationContext(self.annotation), Permission.Annotation.READ
+        # If we can access this without an identity, anyone in the world
+        # can read this. We optimise by checking if we were going to write
+        # "group:__world__" anyway, to save the permission check.
+        if self.annotation.groupid == "__world__" or identity_permits(
+            identity=None,
+            context=AnnotationContext(self.annotation),
+            permission=Permission.Annotation.READ,
```

Comment:
The short cut works in 100% of cases at this point. If your group is `__world__` both of the following paths will write out the same value?

I can split it up a bit more if you think it's clearer?

Reply:
Right but I think the fact that all paths lead to `group:__world__` is because the annotation is known to be shared. This is not true _regardless_ of whether the annotation is shared or not. It's true _because_ the annotation is definitely a shared one.

A non-shared (i.e. private) annotation is only readable by that annotation's creator even if the annotation is in the `__world__` (Public) group or another world readable group. But because the method starts out by doing this:

```python
if not self.annotation.shared:
    return self.annotation.userid

...
```

it gets the private annotations out of the way at the start, so then the following code doesn't have to worry about private annotations in the `__world__` or in another world-readable group.

But the suggestion wasn't about changing what the code actually _does_ but just about making it more readable by letting the structure of the code bring out the logic more. Partly I think having three separate simple `if`'s is easier to read than two `if`'s but one of them is a compound `if` over multiple lines. And partly it's because it brings out the three separate cases. One thing I think it'd be good for reading the code to answer is _What are the different cases?_ that go through this function. What are the different kinds of annotation? In this version I think it's clear that there are four cases:

```
if not self.annotation.shared:
    # Case 1: all private annotations (it doesn't matter what group they're in).
    return self.annotation.userid 

if self.annotation.groupid == "__world__":
    # Case 2: shared annotations in the `__world__` group.
    return "group:__world__"

if identity_permits(identity=None, context=AnnotationContext(self.annotation), permission=Permission.Annotation.READ):
    # Case 3: shared annotations in groups that anyone can read.
    return "group:__world__"

# Case 4: shared annotations in private groups.
return "group:{}".format(self.annotation.groupid)
```

Whereas the current code mixes cases 2 and 3 together with a compound `if`:

```python
if self.annotation.groupid == "__world__" or identity_permits(
    identity=None,
    context=AnnotationContext(self.annotation),
    permission=Permission.Annotation.READ,
):
    # A shared annotation in either the __world__ group or another group that anyone can read.
    return "group:__world__"
```

I think the clue here is when the code comment needs to be a two-line sentence with an "or" in it:

```
# If the group is the public group, or an unauthorized person could
# read from it, it's world readable.
```

The two separate cases conjoined by this "or" can be expressed by the structure of the code instead.